rush-plugins/rush-serve-plugin/src/phasedCommandHandler.ts

```diff
@@ -0,0 +1,131 @@
+      }
+
+      const certManager: CertificateManager = new CertificateManager();
+
+      const certificate: ICertificate = await certManager.ensureCertificateAsync(
+        true, // Allow generating a new certificate
+        logger.terminal
+      );
+
+      const { customParameters } = context;
+
+      let requestedPort: number = 0;
+
+      const portParameter: CommandLineStringParameter | undefined = portParameterLongName
+        ? (customParameters.get(portParameterLongName) as CommandLineStringParameter)
+        : undefined;
+      if (portParameter) {
+        const rawValue: string | undefined = portParameter.value;
+        const parsedValue: number = rawValue ? parseInt(rawValue, 10) : 0;
+        if (!isNaN(parsedValue)) {
+          requestedPort = parsedValue;
+        }
+      }
+
+      const app: express.Express = express();
+
+      // Configure server based on project-level configs
+      // TODO: Update this to be the set of projects selected for the command
+      // Requires a hook to which the project selection is a parameter, minimally
+      const selectedProjects: ReadonlySet<RushConfigurationProject> = new Set(rushConfiguration.projects);
+
+      const serveConfig: RushServeConfiguration = new RushServeConfiguration();
+
+      const routingRules: Iterable<IRoutingRule> = await serveConfig.loadProjectConfigsAsync(
+        selectedProjects,
+        logger.terminal
+      );
+
+      function setHeaders(response: express.Response, path: string, stat: unknown): void {
+        response.set('Access-Control-Allow-Origin', '*');
+        response.set('Access-Control-Allow-Methods', 'GET');
+      }
+
+      for (const rule of routingRules) {
+        app.use(
+          rule.servePath,
+          express.static(rule.diskPath, {
+            dotfiles: 'ignore',
+            immutable: rule.immutable,
+            index: false,
+            redirect: false,
+            setHeaders
+          })
+        );
+      }
+
+      const server: https.Server = https.createServer(
+        {
+          cert: certificate.pemCertificate,
+          key: certificate.pemKey
+        },
+        app
+      );
+
+      server.listen(requestedPort);
+      await once(server, 'listening');
+
+      const address: AddressInfo | undefined = server.address() as AddressInfo;
+      activePort = address?.port;
+
+      logger.terminal.writeLine(`Content is being served from:\n  https://localhost:${activePort}/`);
+      if (portParameter) {
+        // Hack the value of the parsed command line parameter to the actual runtime value.
+        // This will cause the resolved value to be forwarded to operations that may use it.
+        (portParameter as unknown as { _value?: string })._value = `${activePort}`;
+      }
+
+      return operations;
+    }
+  );
+
+  command.hooks.waitingForChanges.tap(PLUGIN_NAME, () => {
+    logger.terminal.writeLine(`Content is being served from:\n  https://localhost:${activePort}/`);
```

Comment:
So will we reprint this whenever Rush has finished rebuilding something? Is that actually useful?

Reply:
The purpose of this hook is to make this information available whenever Rush is paused waiting for file changes, so that the user doesn't have to hunt for it.